# Google Colab Set Up

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Change the directory to your folder on Google Drive

In [ ]:
cd '/content/drive/My Drive/Colab Notebooks/MKTG6010'

/content/drive/My Drive/Colab Notebooks/MKTG6010


In [ ]:
## display the directory
!pwd

/content/drive/My Drive/Colab Notebooks/MKTG6010


# Installing & Importing Packages

## PyLDAVis

### Colab Users

In [ ]:
# Colab users, please install
!pip install PyLDAVis==3.4.1

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Using cached numpy-1.24.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (17.3 MB)
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
tensorflow 2.12.0 requires numpy<1.24,>=1.22, but you have numpy 1.24.3 which is incompatible.
numba 0.56.4 requires numpy<1.24,>=1.18, but you have numpy 1.24.3 which is incompatible.
google-colab 1.0.0 requires pandas~=1.5.3, but you have pandas 2.0.1 which is incompatible.


### Jupyter Users

In [1]:
#Jupyter users, please install
!pip install PyLDAVis==3.4.0

## Import Packages

In [2]:
!pip install numpy==1.8

  Using cached numpy-1.8.0.zip (4.3 MB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'


  error: subprocess-exited-with-error
  
  python setup.py egg_info did not run successfully.
  exit code: 1
  
  [198 lines of output]
  Running from numpy source directory.
  C:\Users\Furqan Ali\AppData\Local\Temp\pip-install-grprk15s\numpy_926e6dc5fcf64c8280c14b173551c684\numpy\distutils\misc_util.py:415: SyntaxWarning: "is" with a literal. Did you mean "=="?
    return is_string(s) and ('*' in s or '?' is s)
  'svnversion' is not recognized as an internal or external command,
  operable program or batch file.
  non-existing path in 'numpy\\distutils': 'site.cfg'
  F2PY Version 2
  blas_opt_info:
  blas_mkl_info:
    libraries mkl,vml,guide not found in ['D:\\SOFTWARE INSTALLED\\ANACONDA\\lib', 'C:\\', 'D:\\SOFTWARE INSTALLED\\ANACONDA\\libs']
    NOT AVAILABLE
  
  openblas_info:
    libraries openblas not found in ['D:\\SOFTWARE INSTALLED\\ANACONDA\\lib', 'C:\\', 'D:\\SOFTWARE INSTALLED\\ANACONDA\\libs']
    NOT AVAILABLE
  
  atlas_blas_threads_info:
  Setting PTATLAS=ATLAS
    l

In [3]:
pip install numpy --upgrade

Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd
import numpy as np
import nltk
from nltk.stem import WordNetLemmatizer
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation
import pyLDAvis.lda_model

#warnings.filterwarnings('ignore')

In [2]:
# %% [code]

# Data Preparation

## Tweets_sport

In [3]:
# Uploading .csv to dataframe and examining info
data_twt=pd.read_csv('D:/WORK/Fiver/Ldgs/04/part 2/tweets_sport.csv', na_values = ' ')
data_twt.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13523 entries, 0 to 13522
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   id                13523 non-null  float64
 1   text              13523 non-null  object 
 2   event_number      13523 non-null  int64  
 3   player_dummy      13523 non-null  int64  
 4   team_dummy        13523 non-null  int64  
 5   weekday           13523 non-null  int64  
 6   month             13523 non-null  int64  
 7   day               13523 non-null  int64  
 8   time              13523 non-null  object 
 9   user_screen_name  13523 non-null  object 
 10  retweets          13522 non-null  float64
 11  likes             13522 non-null  float64
 12  comments          13522 non-null  float64
 13  Unnamed: 13       1 non-null      object 
dtypes: float64(4), int64(6), object(4)
memory usage: 1.4+ MB


In [4]:
# Check the dataset
data_twt.head()

id                                               text   
0  5.210000e+17  "rt @dariofranchitti: this bathurst race has b...  \
1  5.210000e+17  "rt @wtf1couk: gp3 fans while you're waiting f...   
2  5.210000e+17  "rt @sally_fitz: incredible final laps of #bat...   
3  5.210000e+17  "rt @v8supercars: top 10 - mostert, moffat, pe...   
4  5.210000e+17  "#bathurst1000  i present to you, the fully st...   

   event_number  player_dummy  team_dummy  weekday  month  day   time   
0            11             0           0        7     10   12   7:28  \
1            11             0           0        7     10   12   7:28   
2            11             1           0        7     10   12   7:31   
3            11             0           0        7     10   12   9:04   
4            11             0           0        7     10   12  10:20   

  user_screen_name  retweets  likes  comments Unnamed: 13  
0          007_dot       0.0    0.0       0.0              
1          007_dot       0.0    0.0       0.0         NaN  
2           007cnt       0.0    0.0       0.0         NaN  
3         007pixxx       0.0    0.0       0.0         NaN  
4         007pixxx       0.0    0.0       0.0         NaN

In [5]:
# Describing data
data_twt.describe()

id  event_number  player_dummy    team_dummy       weekday   
count  1.352300e+04  13523.000000  13523.000000  13523.000000  13523.000000  \
mean   5.098869e+17      9.730977      0.102936      0.008726      6.815426   
std    2.537884e+16      3.038715      0.303886      0.093007      0.444136   
min    4.390000e+17      1.000000      0.000000      0.000000      5.000000   
25%    5.210000e+17     11.000000      0.000000      0.000000      7.000000   
50%    5.210000e+17     11.000000      0.000000      0.000000      7.000000   
75%    5.210000e+17     11.000000      0.000000      0.000000      7.000000   
max    5.410000e+17     14.000000      1.000000      1.000000      7.000000   

              month           day      retweets         likes      comments  
count  13523.000000  13523.000000  13522.000000  13522.000000  13522.000000  
mean       8.971974     12.715817      0.338707      0.623281      0.113223  
std        2.292286      5.488871      3.102980      5.086552      0.709985  
min        2.000000      1.000000      0.000000      0.000000      0.000000  
25%       10.000000     12.000000      0.000000      0.000000      0.000000  
50%       10.000000     12.000000      0.000000      0.000000      0.000000  
75%       10.000000     12.000000      0.000000      0.000000      0.000000  
max       12.000000     29.000000    265.000000    303.000000     39.000000

In [6]:
# Replace blanks with null values
data_twt = data_twt.replace(' ', np.nan)

# Confirming null values present in dataframe
data_twt.isnull().sum().sort_values(ascending=False) # Sorted highest to lowest so any columns with nulls will appear at the top

Unnamed: 13         13522
retweets                1
likes                   1
comments                1
id                      0
text                    0
event_number            0
player_dummy            0
team_dummy              0
weekday                 0
month                   0
day                     0
time                    0
user_screen_name        0
dtype: int64

In [7]:
# Replace the null values for numerical fields with mean
cols = ['retweets','likes','comments']
data_twt = data_twt.fillna(value=data_twt[cols].mean())

# Checking there's no nulls left
data_twt.isnull().sum().sort_values(ascending=False)

Unnamed: 13         13522
id                      0
text                    0
event_number            0
player_dummy            0
team_dummy              0
weekday                 0
month                   0
day                     0
time                    0
user_screen_name        0
retweets                0
likes                   0
comments                0
dtype: int64

## Tv_rating

In [8]:
data_rtg=pd.read_csv('D:/WORK/Fiver/Ldgs/04/part 2/tv_rating.csv')
data_rtg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3194 entries, 0 to 3193
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   tv_channel    3194 non-null   object
 1   city          3194 non-null   object
 2   tvviewers     3194 non-null   int64 
 3   start_time    3194 non-null   object
 4   end_time      3194 non-null   object
 5   event_number  3194 non-null   int64 
 6   weekday       3194 non-null   int64 
dtypes: int64(3), object(4)
memory usage: 174.8+ KB


In [9]:
# Check the dataset
data_rtg.head()

tv_channel    city  tvviewers start_time end_time  event_number  weekday
0       ATN7  Sydney         69      14:00    14:14             1        5
1       ATN7  Sydney         56      14:15    14:29             1        5
2       ATN7  Sydney         51      14:30    14:44             1        5
3       ATN7  Sydney         48      14:45    14:59             1        5
4       ATN7  Sydney         48      15:00    15:14             1        5

In [10]:
# Describing data
data_rtg.describe()

tvviewers  event_number      weekday
count  3194.000000   3194.000000  3194.000000
mean     65.638698      7.556356     6.398560
std      75.464908      4.225938     0.606271
min       0.000000      1.000000     5.000000
25%      20.000000      4.000000     6.000000
50%      41.000000      8.000000     6.000000
75%      89.000000     11.000000     7.000000
max     752.000000     14.000000     7.000000

In [11]:
# Replace blanks with null values
data_rtg = data_rtg.replace(' ', np.nan)

# Confirming no null values present in dataframe
data_rtg.isnull().sum().sort_values(ascending=False) # Sorted highest to lowest so any columns with nulls will appear at the top

tv_channel      0
city            0
tvviewers       0
start_time      0
end_time        0
event_number    0
weekday         0
dtype: int64

# Q1: EDA of binary variables (player, team, etc.) of all tweets

In [12]:
# WIP

In [13]:
data_twt.head()

id                                               text   
0  5.210000e+17  "rt @dariofranchitti: this bathurst race has b...  \
1  5.210000e+17  "rt @wtf1couk: gp3 fans while you're waiting f...   
2  5.210000e+17  "rt @sally_fitz: incredible final laps of #bat...   
3  5.210000e+17  "rt @v8supercars: top 10 - mostert, moffat, pe...   
4  5.210000e+17  "#bathurst1000  i present to you, the fully st...   

   event_number  player_dummy  team_dummy  weekday  month  day   time   
0            11             0           0        7     10   12   7:28  \
1            11             0           0        7     10   12   7:28   
2            11             1           0        7     10   12   7:31   
3            11             0           0        7     10   12   9:04   
4            11             0           0        7     10   12  10:20   

  user_screen_name  retweets  likes  comments Unnamed: 13  
0          007_dot       0.0    0.0       0.0              
1          007_dot       0.0    0.0       0.0         NaN  
2           007cnt       0.0    0.0       0.0         NaN  
3         007pixxx       0.0    0.0       0.0         NaN  
4         007pixxx       0.0    0.0       0.0         NaN

In [14]:
data_twt['event_number'].value_counts()

event_number
11    9620
1      603
12     503
4      493
3      344
14     274
10     273
6      265
9      233
5      227
7      217
2      196
8      149
13     126
Name: count, dtype: int64

In [25]:
# data from https://allisonhorst.github.io/palmerpenguins/

import matplotlib.pyplot as plt

event = data_twt['event_number']
event_values = {
    'Bill Depth': (18.35, 18.43, 14.98),
    'Bill Length': (38.79, 48.83, 47.50),
    'Flipper Length': (189.95, 195.82, 217.19),
}

x = np.arange(len(species))  # the label locations
width = 0.25  # the width of the bars
multiplier = 0

fig, ax = plt.subplots(layout='constrained')

for attribute, measurement in penguin_means.items():
    offset = width * multiplier
    rects = ax.bar(x + offset, measurement, width, label=attribute)
    ax.bar_label(rects, padding=3)
    multiplier += 1

# Add some text for labels, title and custom x-axis tick labels, etc.
ax.set_ylabel('Length (mm)')
ax.set_title('Penguin attributes by species')
ax.set_xticks(x + width, species)
ax.legend(loc='upper left', ncols=3)
ax.set_ylim(0, 250)

plt.show()

NameError: name 'species' is not defined

In [26]:
data_rtg.head()

tv_channel    city  tvviewers start_time end_time  event_number  weekday
0       ATN7  Sydney         69      14:00    14:14             1        5
1       ATN7  Sydney         56      14:15    14:29             1        5
2       ATN7  Sydney         51      14:30    14:44             1        5
3       ATN7  Sydney         48      14:45    14:59             1        5
4       ATN7  Sydney         48      15:00    15:14             1        5

# Q2: What are the most common topics discussed by Supercar fans on Twitter during live racing events, and how do these topics relate to online user engagement and live television viewership?

## Topic Modelling

In [27]:
# Isolate to just the tweet text
text=data_twt['text']
# Change the text to lower case
text = text.map(lambda text: text.lower())
# Convert data from pandas format to list values  
text=text.values.tolist()

#print the first 3 tweets
print(text[:3])

['"rt @dariofranchitti: this bathurst race has been an absolute belter!! love the aussie commentary. was hoping for better for my mate @jcourt…"', '"rt @wtf1couk: gp3 fans while you\'re waiting for the race to resume, flick over to the #bathurst1000. exciting final few laps."', '"rt @sally_fitz: incredible final laps of #bathurst1000 remarkable comeback from @jamiewhincup coming from the back of the grid @redbullraci…"']


In [28]:
# Download wordnet
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to C:\Users\Furqan
[nltk_data]     Ali\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [29]:
# Conduct lemmatization for the words in the text
tokens=[]
for sent in text:
    temp=[WordNetLemmatizer().lemmatize(word) for word in sent.split(" ")]
    tokens.append(temp)

In [30]:
# Customized the stopwords
from sklearn.feature_extraction import text # import cannot be moved to top or error occurs
my_additional_stop_words = ["bathurst","adelaide","http","tas","tcm","darwin","coateshire","sandown","clipsal","sydney", "townsville"]
stop_words = text.ENGLISH_STOP_WORDS.union(my_additional_stop_words)

### Latent Dirichlet Allocation

In [31]:
%%time

# Set how many topics we need
NUM_TOPICS = 10
data_twt['tokens']=tokens
text_train = list(data_twt['tokens'].apply(lambda x: ' '.join(x)))

# Convert a collection of text documents to a matrix of token counts
vectoriser = CountVectorizer(min_df=5, max_df=0.9,
                             stop_words=list(stop_words),lowercase=True, # Updated from tutorial code
                             token_pattern='[a-zA-Z\-][a-zA-Z\-]{2,}')

# Fit
data_vectorized=vectoriser.fit_transform(text_train)

# Build a Latent Dirichlet Allocation Model
lda_model = LatentDirichletAllocation(n_components=NUM_TOPICS, max_iter=10, random_state=258, learning_method='online')
lda_Z = lda_model.fit_transform(data_vectorized)


<unknown>:9: DeprecationWarning: invalid escape sequence \-


CPU times: total: 27.3 s
Wall time: 27.6 s


In [32]:
# Print shape
print(lda_Z.shape)

(13523, 10)


In [33]:
# Getting top 10 words per topic
def print_topics(model, vectoriser, top_n=10):
    for idx, topic in enumerate(model.components_):
        print("Topic %d:" % (idx+1))
        print([(vectoriser.get_feature_names_out()[i], topic[i]) # Updated from tutorial code
                        for i in topic.argsort()[:-top_n - 1:-1]])
 
print("LDA Model:")
print_topics(lda_model, vectoriser)

LDA Model:
Topic 1:
[('morris', 607.0090904416668), ('chaz', 579.8046379508434), ('mostert', 542.3464781221942), ('paul', 531.4341916057526), ('just', 513.6636136315058), ('supercars', 431.8760032778609), ('mountain', 361.80821075336416), ('wow', 301.0098232730656), ('king', 268.26003200918177), ('xwrcjhmg', 206.1980304156728)]
Topic 2:
[('amp', 413.73167949706755), ('like', 327.6511433162216), ('day', 290.8138715867198), ('race', 148.34486606794226), ('start', 146.49978746828526), ('believe', 143.1790926982608), ('car', 127.8189733117516), ('thanks', 117.68380685793879), ('home', 116.12669092790851), ('job', 115.75436904959923)]
Topic 3:
[('car', 536.1150617409814), ('supercars', 370.24481507630844), ('fuel', 323.79145578324295), ('year', 256.2866649449783), ('whincup', 225.08046770848148), ('sport', 197.91005371165315), ('make', 192.33190699572577), ('safety', 177.19981686529334), ('svg', 171.29447437734493), ('got', 149.81890134761593)]
Topic 4:
[('amazing', 331.4420169949792), ('wa

In [34]:
# Generating topic fields in df
x = lda_model.transform(data_vectorized)

topics=pd.DataFrame(x)
topics.columns=['topic1','topic2','topic3','topic4','topic5','topic6','topic7','topic8','topic9','topic10']

topics['tweet']=data_twt['text']
topics['month']=data_twt['month']
topics['day']=data_twt['day']
pd.DataFrame(topics).to_csv('D:/WORK/Fiver/Ldgs/04/part 2/tweets_by_topics_all_new.csv',index=False)

In [35]:
topics

topic1    topic2    topic3    topic4    topic5    topic6    topic7   
0      0.008333  0.008333  0.008333  0.008333  0.008333  0.924999  0.008333  \
1      0.009091  0.009093  0.009095  0.009091  0.009092  0.009095  0.827256   
2      0.008333  0.008333  0.008334  0.008334  0.008333  0.008333  0.008333   
3      0.008334  0.008333  0.008334  0.008333  0.008333  0.008333  0.008334   
4      0.025000  0.025000  0.025000  0.275001  0.025000  0.025015  0.025000   
...         ...       ...       ...       ...       ...       ...       ...   
13518  0.016667  0.016667  0.016667  0.183333  0.016667  0.016670  0.683330   
13519  0.050000  0.050000  0.549994  0.050000  0.050000  0.050000  0.050000   
13520  0.025004  0.274993  0.025000  0.525000  0.025000  0.025000  0.025000   
13521  0.016667  0.183332  0.016667  0.016667  0.183334  0.016671  0.349997   
13522  0.033333  0.033333  0.700000  0.033333  0.033333  0.033333  0.033333   

         topic8    topic9   topic10   
0      0.008333  0.008334  0.008333  \
1      0.100000  0.009097  0.009091   
2      0.175000  0.758332  0.008333   
3      0.924998  0.008333  0.008333   
4      0.524984  0.025000  0.025000   
...         ...       ...       ...   
13518  0.016667  0.016667  0.016667   
13519  0.050006  0.050000  0.050000   
13520  0.025000  0.025003  0.025000   
13521  0.016667  0.016667  0.183332   
13522  0.033333  0.033333  0.033333   

                                                   tweet  month  day  
0      "rt @dariofranchitti: this bathurst race has b...     10   12  
1      "rt @wtf1couk: gp3 fans while you're waiting f...     10   12  
2      "rt @sally_fitz: incredible final laps of #bat...     10   12  
3      "rt @v8supercars: top 10 - mostert, moffat, pe...     10   12  
4      "#bathurst1000  i present to you, the fully st...     10   12  
...                                                  ...    ...  ...  
13518  "oh, that is absolutely heartbreaking for shan...     10   12  
13519  "rt @off1cial_sport: clipsal 500 #adelaide htt...      3    1  
13520  "storm &amp; sunset #adelaide http://t.co/8gym...     10   26  
13521  ""you're a pack of arseholes" still the greate...     10   12  
13522  "rt @nathanepic: omg so much drama :p #bathurs...     10   12  

[13523 rows x 13 columns]

## Topic Map

In [ ]:
pyLDAvis.enable_notebook()

# Creating panel map for detailed topic and keyword examination
panel = pyLDAvis.lda_model.prepare(lda_model, data_vectorized, vectoriser, sort_topics = False)
panel

# Q3: How does the sentiment of live tweets during Supercar racing events impact online user engagement and live television viewership?

The sentiment of live tweets during Supercar racing events can impact online user engagement and live television viewership in several ways. Positive sentiment expressed on social media can increase the excitement and enthusiasm for the event, which in turn can encourage more people to watch and engage with the event. Positive sentiment can also lead to increased sharing and discussion of content related to the event.
On the other hand, negative sentiment expressed on social media can have a dampening effect on engagement and viewership. Negative sentiment can discourage people from watching or engaging with the event and can lead to decreased sharing and discussion of content related to the event.

In [16]:
import pandas as pd
import nltk
from nltk.sentiment.vader import SentimentIntensityAnalyzer


In [17]:
tweet_data = pd.read_csv('D:/WORK/Fiver/Ldgs/04/part 2/tweets_sport.csv')
tv_rating_data = pd.read_csv('D:/WORK/Fiver/Ldgs/04/part 2/tv_rating.csv')


In [18]:
nltk.download('vader_lexicon')
sid = SentimentIntensityAnalyzer()

tweet_data['sentiment'] = tweet_data['text'].apply(lambda x: sid.polarity_scores(x)['compound'])


[nltk_data] Downloading package vader_lexicon to C:\Users\Furqan
[nltk_data]     Ali\AppData\Roaming\nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [19]:

tweet_data['time'] = pd.to_datetime(tweet_data['time'])
tweet_data.set_index('time', inplace=True)

C:\Users\Furqan Ali\AppData\Local\Temp\ipykernel_4052\1671028379.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  tweet_data['time'] = pd.to_datetime(tweet_data['time'])


In [20]:
# Convert the sentiment column to a numeric type
tweet_data['sentiment'] = pd.to_numeric(tweet_data['sentiment'], errors='coerce')

# Group the data by minute and calculate the mean sentiment and engagement
tweet_data_grouped = tweet_data.resample('1min').agg({'sentiment': 'mean', 'retweets': 'sum', 'likes': 'sum'})
tweet_data_grouped['engagement'] = tweet_data_grouped[['retweets', 'likes']].sum(axis=1)



In [21]:
tv_rating_data['start_time'] = pd.to_datetime(tv_rating_data['start_time'])
tv_rating_data.set_index('start_time', inplace=True)
tv_rating_data['end_time'] = pd.to_datetime(tv_rating_data['end_time'])
tv_rating_data.set_index('end_time', inplace=True)


C:\Users\Furqan Ali\AppData\Local\Temp\ipykernel_4052\891056771.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  tv_rating_data['start_time'] = pd.to_datetime(tv_rating_data['start_time'])
C:\Users\Furqan Ali\AppData\Local\Temp\ipykernel_4052\891056771.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  tv_rating_data['end_time'] = pd.to_datetime(tv_rating_data['end_time'])


In [22]:
merged_data = tweet_data_grouped.join(tv_rating_data, how='outer')
merged_data.dropna(inplace=True)

In [23]:
corr_sentiment_engagement = merged_data[['sentiment', 'engagement']].corr().iloc[0, 1]
corr_sentiment_ratings = merged_data[['sentiment', 'tvviewers']].corr().iloc[0, 1]


In [24]:
print("Correlation between sentiment and engagement:", corr_sentiment_engagement)
print("Correlation between sentiment and TV ratings:", corr_sentiment_ratings)


Correlation between sentiment and engagement: -0.39085867008626496
Correlation between sentiment and TV ratings: 0.01360700489625919


### results
we had performed data aggregation and correlation analysis on two data sources: tweet_sport and Tv_rating.we aggregates tweet data by minute and calculates the mean sentiment and engagement, then calculates the correlation between sentiment and engagement, and sentiment and TV ratings, which shows us the relation between the tweets and tv rating.

Important steps followed:

•	The sentiment of each tweet is calculated using the sentiment analysis tool.<br>
•	The polarity_scores function is used to calculate the sentiment score for each tweet.<br>
•	The time column in the tweet data is converted to a pandas datetime format and set as the index for the data frame.<br>
•	The sentiment column is converted to a numeric type so that it can be used for calculations later.<br>
•	The tweet data is then resampled by minute and aggregated to calculate the mean sentiment and engagement (sum of retweets and likes) for each minute.<br>
•	The TV rating data is similarly formatted by converting the start and end times to datetime format and setting them as the index for the data frame.<br>
•	The tweet data and TV rating data are merged together using an outer join, so that all data is included even if there are missing values.<br>
•	Any rows with missing values are dropped from the merged data frame.
•	The correlation between sentiment and engagement, as well as between sentiment and TV ratings, is calculated using the corr() function.<br>



The output shows the correlation coefficients between the sentiment of tweets and two other variables: engagement (measured as the sum of retweets and likes) and TV ratings.<br>
    
•	The correlation coefficient between sentiment and engagement is -0.39, which indicates a moderate negative correlation. This means that as the sentiment of tweets becomes more negative, the level of engagement (retweets and likes) tends to decrease.<br>
•	The correlation coefficient between sentiment and TV ratings is 0.01, which indicates a very weak positive correlation. This means that there is little to no relationship between the sentiment of tweets and TV ratings.<br>

### Limitation :

This analysis has some limitations that should be considered when interpreting the results.<br>
Firstly, the sentiment analysis algorithm used in this analysis, the VADER algorithm, may not be entirely accurate in assessing the sentiment of tweets. The VADER algorithm is based on a lexicon of words that have been manually labeled with their sentiment score, which may not always capture the nuances and context of language used in tweets. Additionally, the sentiment score assigned to a tweet is based on a set of predetermined rules and heuristics, which may not always be applicable to different types of tweets or topics.
<br>Secondly, the sample of tweets used in this analysis may not be representative of the entire population of tweets related to sports. The tweets were obtained from a single source, and only include tweets that contain certain keywords related to sports. This may lead to a biased sample that does not accurately reflect the sentiment or engagement of the broader population of tweets related to sports.
<br>Thirdly, the analysis assumes that there is a causal relationship between sentiment and engagement or TV ratings. While there may be a correlation between these variables, it is not necessarily the case that changes in sentiment directly cause changes in engagement or TV ratings. Other factors, such as the timing of events or external news events, may also influence engagement or TV ratings and need to be accounted for in the analysis.

<br>The analysis in question 3 only considers correlations between variables and does not establish any causal relationships.



# Q4: Can we predict online user engagement and live television viewership for Supercar racing events based on the content and sentiment of live tweets?

it is possible to use the content and sentiment of live tweets to make predictions about online user engagement and live television viewership for Supercar racing events, but the accuracy of these predictions may depend on various factors.
Quality:
    The quality of the sentiment analysis algorithm used to analyze the tweets will be crucial in determining the accuracy of 
    the predictions.

Size and Representativeness of sample:
    The size and representativeness of the sample of tweets used to make predictions will also be important

other factors such as the time of day, day of the week, and the specific Supercar racing event being analyzed may also impact the accuracy of the predictions.

# Q5: What is the relationship between the time of day when tweets are posted and the online user engagement and live television viewership for Supercar racing events?

The relationship between the time of day when tweets are posted and the online user engagement and live television viewership for Supercar racing events can vary depending on several factors.

For example, if the Supercar racing event is happening in a different time zone than the majority of its viewers, then the time of day when the event is happening may not align with peak engagement times for those viewers. In this case, tweets may still generate engagement, but it may not have as much impact on live television viewership.

On the other hand, if the Supercar racing event is happening during peak viewing hours for its target audience, then tweets posted during that time may have a stronger impact on both online user engagement and live television viewership.